internal/pkg/earpointer/analysis.go

```diff
@@ -0,0 +1,649 @@
+	vis.processHeapAccess(send.X, send.Chan, anyIndexField)
+}
+
+func (vis *visitor) visitRange(rng *ssa.Range) {
+	// t1 = range t0: unify t0 and t1.
+	// It will be processed by the subsequent extract instruction that also handles the relevant Next.
+}
+
+func (vis *visitor) VisitSelect(s *ssa.Select) {
+	// TODO: to be implemented.
+}
+
+func (vis *visitor) visitCall(call ssa.CallCommon, instr ssa.Instruction) {
+	// TODO: to be added.
+}
+
+// Process a load/store using an index (which can be constant).
+func (vis *visitor) processMemoryIndexAccess(data ssa.Value, base ssa.Value, index ssa.Value) {
+	// If the index is a constant, its string name is used as the field name;
+	// otherwise the predefined field name "AnyField" is used.
+	var field Field
+	if c, ok := index.(*ssa.Const); ok {
+		field = Field{Name: c.Value.String()}
+	} else {
+		field = anyIndexField
+	}
+	vis.processHeapAccess(data, base, field)
+}
+
+// Process a load/store from data_reg to (base, index), where base can be a register
+// or global variable. Here heap access is through unify-by-reference simulation.
+func (vis *visitor) processHeapAccess(data ssa.Value, base ssa.Value, fd Field) {
+	// Skip unification if lhs is not a sharable type.
+	if !typeMayShareObject(data.Type()) {
+		return
+	}
+	// unify instance field
+	for _, context := range vis.contexts {
+		vis.unifyField(context, base, fd, data)
+	}
+}
+
+// Process memory operator & and * by making the address pointing to the value.
+func (vis *visitor) processAddressToValue(addr ssa.Value, value ssa.Value) {
+	if !typeMayShareObject(addr.Type()) || !typeMayShareObject(value.Type()) {
+		return
+	}
+	state := vis.state
+	for _, context := range vis.contexts {
+		caddr := state.Insert(MakeReference(context, addr))
+		cvalue := state.Insert(MakeReference(context, value))
+		fmap := state.PartitionFieldMap(state.representative(caddr))
+		// Add or unify the field.
+		if v, ok := fmap[directPointToField]; ok {
+			if isUnifyByReference(value.Type()) { // unify-by-reference
+				state.Unify(v, cvalue)
+			} else {
+				vis.unifyByValue(v, cvalue)
+			}
+		} else {
+			fmap[directPointToField] = cvalue
+		}
+	}
+}
+
+// Unify local regs "v1" and "v2" in all contexts.
+func (vis *visitor) unifyLocals(v1 ssa.Value, v2 ssa.Value) {
+	if !mayShareObject(v1) || !mayShareObject(v2) {
+		return
+	}
+	state := vis.state
+	for _, context := range vis.contexts {
+		state.Unify(MakeReference(context, v1), MakeReference(context, v2))
+	}
+}
+
+// Unify "obj.field" and "target" in "context".
+func (vis *visitor) unifyField(context *Context, obj ssa.Value, fd Field, target ssa.Value) {
+	if !typeMayShareObject(target.Type()) {
+		return
+	}
+	// This generates the unification constraint obj.field = target
+	state := vis.state
+	objr := MakeReference(context, obj)
+	fmap := state.PartitionFieldMap(state.representative(objr))
+	if fmap == nil { // Should not happen
+		log.Fatal("field map is nil")
```

Comment:
I don't really like a `log.Fatal`s from somewhere deep in program logic.  If this is a Catastrophic, All Results Are Now Invalid sort of error, then a hard-exit is probably okay.

In either case, the message provided to a user should be something that is actionable, e.g. the request for a report like that in `levee.go:123`:
```
fmt.Printf("unexpected node received: %v (type %T); please report this issue\n", path[0], path[0])
```

----

Ditto the three other `log.Fatal` calls.